framework/py/flwr/supernode/nodestate/nodestate.py

```
@@ -168,3 +168,39 @@ def get_run_ids_with_pending_messages(self) -> Sequence[int]:
         Sequence[int]
             Sequence of run IDs with pending messages.
         """
+
+    @abstractmethod
+    def record_message_processing_start(self, message_id: str) -> None:
+        """Record the start time of message processing based on the message ID.
+
+        Parameters
+        ----------
+        message_id : str
+            The ID of the message associated with the start time.
+        """
+
+    @abstractmethod
+    def record_message_processing_end(self, message_id: str) -> None:
+        """Record the end time of message processing based on the message ID.
+
+        Parameters
+        ----------
+        message_id : str
+            The ID of the message associated with the end time.
+        """
+
+    @abstractmethod
+    def get_message_processing_duration(self, message_id: str) -> float | None:
+        """Get the message processing duration based on the message ID.
+
+        Parameters
+        ----------
+        message_id : str
+            The ID of the message.
+
+        Returns
+        -------
+        Optional[float]
+            The processing duration in seconds. Returns None if message ID not found,
+            or if start/end times are missing.
```

Hmmm, should we raise a ValueError when the message ID is missing? I mean, it seems that in `record_message_processing_end`, we will raise an exception when the ID is missing, right?